Follow Me mode Second order filter & improvements

junwoo091400 opened the pull request:
**Describe problem solved by this pull request**
Refer to the document : https://docs.google.com/document/d/1E04Rn6sxYNqKve4-KZYdC2-9LExKmHKuYjWZlwf61rA/edit

**Describe your solution**

**Describe possible alternatives**
A clear and concise description of alternative solutions or features you've considered.

**Test data / coverage**
This change log : https://logs.px4.io/plot_app?log=6561ce69-e2e3-46c3-9e13-e29f2dbc11c7
Alessandro PR log : https://logs.px4.io/plot_app?log=fa76985d-87c6-4c4f-9164-fa2f2c15338b

**Additional context**
* This PR is to be merged to a pre-existing PR : #18026 
* Depends on  library added from #19246 and #19249.
* Proposal brainstorming issue : #19251 
* User Doc Update PR : https://github.com/PX4/PX4-user_guide/pull/1810
* Scripts I used for testing the algorithm : https://github.com/junwoo091400/Follow-Target-Algorithm-Scripts
* Script used to test Follow-Me in SITL : https://github.com/junwoo091400/PX4-Followme-Sim-Script
* Visualizer script that was used for drone's orbit angle set-point overshooting problem : https://github.com/junwoo091400/Follow-Target-uLog-Visualizer

### Discussions
1. Should we just change the 'follow perspective' into a single number parameter (0 ~ 360 degrees), instead of using an enum / selection of : "Front, Back, Right, Left, etc."?
2. Google Astyle for the member variables : Do we need to add the _ suffix?

### TODOS
- [x] Reduce follow_target_status logging rate to 200ms interval or 400ms
- [ ] Later on implement the Multcopter Rate Controller not directly applying acceleration setpoint -> Issue at #19543 
- [ ] Maybe move onto [YAML based](https://docs.px4.io/master/en/advanced/parameters_and_configurations.html#parameter-metadata) Parameter definitions?
- [ ] MAVSDK Param update
  - https://github.com/mavlink/MAVSDK-Proto/pull/287
  - 
- [ ] Add the Leash mode (as if drone was hovering and the user was connected to the drone, so it would follow around like as if it was leashed)
- [ ] Add fixed orbit angle mode (staying always on the north side, etc. No orbit angle transitions)
- [ ] [Check if I get a more stable yaw setpoint if I use the heading from the position setpoint to the target estimate, instead of the real position.](https://github.com/PX4/PX4-Autopilot/pull/19260#discussion_r869034921)

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -170,8 +174,8 @@ class FlightTaskAutoFollowTarget : public FlightTask
-	// Lowpass filters for smoothingtarget position because it's used for setpoint generation
-	AlphaFilter<matrix::Vector3f> _target_position_filter;
+	// Second Order Filter to calculate kinematically feasible target position
+	SecondOrderReferenceModel<matrix::Vector3f> _target_position_filter;
```
I would recommend naming it `_target_pose_filter`, as it is not just filtering position, but also velocity.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -187,22 +189,16 @@ Vector3f FlightTaskAutoFollowTarget::calculate_drone_desired_position(Vector3f t
 
-Vector3f FlightTaskAutoFollowTarget::calculate_target_position_filtered(Vector3f pos_ned_est, Vector3f vel_ned_est,
-		Vector3f acc_ned_est)
+Vector3f FlightTaskAutoFollowTarget::calculate_target_position_filtered(Vector3f pos_ned_est, Vector3f vel_ned_est)
```
Similar comment on position vs pose - maybe the method could be called `update_filtered_target_pose()` ? Here note you probably don't need to return anything.. just `.getState()` and `.getRate()` each time further down in the code you need the filtered position or velocity. The vector is already stored once there, no need for a copy.

On another note.. if there is only the reset check happening in this method, would also be valid to move it out into the main `update()` function.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -187,22 +189,16 @@ Vector3f FlightTaskAutoFollowTarget::calculate_drone_desired_position(Vector3f t
 	// Reset the smoothness filter once the target position estimate is available
 	if (!PX4_ISFINITE(_target_position_filter.getState()(0)) || !PX4_ISFINITE(_target_position_filter.getState()(1))
 	    || !PX4_ISFINITE(_target_position_filter.getState()(2))) {
 		_target_position_filter.reset(pos_ned_est);
```
Filter will converge faster if the `vel_ned_est` is also reset here.

Also - we may want to add the velocity components of the filter to this check .. in case any of them became !finite.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -187,22 +189,16 @@ Vector3f FlightTaskAutoFollowTarget::calculate_drone_desired_position(Vector3f t
+	// Second order target position filter to calculate kinematically feasible target position
+	_target_position_filter.update(_deltatime, pos_ned_est, vel_ned_est);
 
 	return _target_position_filter.getState();
```
would remove the return -- see comment from above

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -91,6 +92,10 @@ static constexpr float DIRECTION_FILTER_ALPHA =	3.0f;
 static constexpr float VELOCITY_FF_FILTER_ALPHA = 1.0f;
 
+// Second order filter parameter for target position filter
+static constexpr float TARGET_POSITION_FILTER_NATURAL_FREQUENCY = 4.0f; // [rad/s]
+static constexpr float TARGET_POSITION_FILTER_DAMPING_RATIO = 0.7071;
```
similar naming comment - would use `pose`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -154,8 +159,7 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	matrix::Vector2f calculate_offset_vector_filtered(matrix::Vector3f vel_ned_est);
-	matrix::Vector3f calculate_target_position_filtered(matrix::Vector3f pos_ned_est, matrix::Vector3f vel_ned_est,
-			matrix::Vector3f acc_ned_est);
+	matrix::Vector3f calculate_target_position_filtered(matrix::Vector3f pos_ned_est, matrix::Vector3f vel_ned_est);
```
additionally on naming .. this is more a pedantic request.. but it is nice when method names are not spoken like yoda :) e.g. `update_filtered_target_pose`

junwoo091400 replied:
I will keep Yoda away ;)

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -270,37 +327,35 @@ bool FlightTaskAutoFollowTarget::update()
-			break;
+				// Unwrap : Needed since when filter's tracked state is around -M_PI, and the raw angle goes to
+				// +M_PI, the filter can just average them out and give wrong output.
+				float yaw_setpoint_raw_unwrapped = matrix::unwrap_pi(_yaw_setpoint_filter.getState(), _drone_to_target_heading);
```
```suggestion
				const float yaw_setpoint_raw_unwrapped = matrix::unwrap_pi(_yaw_setpoint_filter.getState(), _drone_to_target_heading);
```

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
 	// Z-position based off curent and initial target altitude
-	// TODO: Parameter NAV_MIN_FT_HT has been repurposed to be used as the desired
-	// altitude above the target. Clarify best solution.
 	switch (_param_nav_ft_alt_m.get()) {
```
also, could we make this more constrained to specifically the mode enum type? i.e. define a name for the altitude modes enum struct (see comment in header) and then pass this enum type into this method.

e.g.
```
Vector3f FlightTaskAutoFollowTarget::calculate_desired_drone_position(const Vector3f &target_position, const FollowAltitudeMode &follow_altitude_mode)
```

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -257,9 +317,6 @@ bool FlightTaskAutoFollowTarget::update()
-		_velocity_ff_scale.update(desired_velocity_ff_scale);
-
 		// Emergency ascent when too close to the ground
 		_emergency_ascent = PX4_ISFINITE(_dist_to_ground) && _dist_to_ground < MINIMUM_SAFETY_ALTITUDE;
```
why does this need to be an internal state?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	// Tracked orbit tangential speed, to compensate for the orbital motion for velocity setpoints
+	Vector2f _orbit_tangential_velocity{0.0f, 0.0f};
 
 	// NOTE: If more of these internal state variables come into existence, it
 	// would make sense to create an internal state machine with a single enum
 	bool _emergency_ascent = false;
```
already commented as well in cpp file .. but again - why does this need to be an internal state? can we delete?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -309,18 +364,23 @@ bool FlightTaskAutoFollowTarget::update()
 }
 
-float FlightTaskAutoFollowTarget::update_follow_me_angle_setting(int param_nav_ft_fs) const
+float FlightTaskAutoFollowTarget::get_follow_me_angle_setting(int param_nav_ft_fs) const
```
can we make this more precise and demand the specific enum type? e.g. `const FollowPerspective &follow_perspective`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -359,13 +419,6 @@ float FlightTaskAutoFollowTarget::update_follow_me_angle_setting(int param_nav_f
-}
-
-
 void FlightTaskAutoFollowTarget::point_gimbal_at(float xy_distance, float z_distance)
```
```suggestion
void FlightTaskAutoFollowTarget::point_gimbal_at(const float xy_distance, const float z_distance)
```

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -72,26 +70,38 @@ static constexpr float ALT_ACCEPTANCE_THRESHOLD = 3.0f;
 // is too close to the ground (below MINIMUM_SAFETY_ALTITUDE)
 static constexpr float EMERGENCY_ASCENT_SPEED = 0.2f;
```
is this a bit slow? what if velocity tracking is a bit bad? may not even go up. this is also after all an *emergency* ascent. @potaito, wdyt?

junwoo091400 replied:
Updated to 0.5!

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -72,26 +70,38 @@ static constexpr float ALT_ACCEPTANCE_THRESHOLD = 3.0f;
 // is too close to the ground (below MINIMUM_SAFETY_ALTITUDE)
```
also need units here

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -72,26 +70,38 @@ static constexpr float ALT_ACCEPTANCE_THRESHOLD = 3.0f;
-// Lowpass applied to the actual NED direction how the drone is facing the target
-// regarless of the setting. Used for dynamic tracking angles when the target makes a turn
-static constexpr float DIRECTION_FILTER_ALPHA =	3.0f;
+// [m/s] Velocity deadzone for which, under this velocity, the target orientation
```
*speed

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -72,26 +70,38 @@ static constexpr float ALT_ACCEPTANCE_THRESHOLD = 3.0f;
+// [m/s] Velocity limit to limit orbital angular rate depending on follow distance
+static constexpr float MAXIMUM_TANGENTIAL_ORBITING_SPEED = 5.0;
+
+// Yaw setpoint filter to avoid jitter-ness, which can happen because the yaw is
```
units

Jaeyoung-Lim commented:
@junwoo091400 Probably the flightgear bridge submodule has been updated unintentionally. I don't think this should be part of this PR

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -104,6 +114,7 @@ class FlightTaskAutoFollowTarget : public FlightTask
 
 protected:
+	// Follow Perspectives set by the parameter NAV_FT_FS
 	enum {
```
Let's give this a name -- e.g. `FollowPerspective`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -104,6 +114,7 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	bool update() override;
 
 protected:
+	// Follow Perspectives set by the parameter NAV_FT_FS
```
you need to check that this 0-360 definition still works with the -pi,pi definition that exists everywhere else in your logic now. otherwise we should redefine this to -180 to 180 for follow perspective (i anyway think this is a bit more intuitive)

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	};
 
+	// Follow Altitude modes set by the parameter NAV_FT_ALT_M
 	enum {
```
let's give this a name -- e.g. `FollowAltitudeMode`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	 *
-	 * @param param_nav_ft_fs value of the parameter NAV_FT_FS
-	 * @return Angle [deg] from which the drone should view the target while following it, with zero degrees indicating the target's 12 o'clock
+	 * @return Angle [rad] Tracked target orientation (heading)
```
also need `@param[in]` description.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	/**
-	 * Predict target's position through forward integration of its currently estimated position, velocity and acceleration.
+	 * Updates the orbit angle setpoint, taking into account the maximal orbit tangential speed
 	 *
```
`@param[in]`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
 
+	/**
+	 * Calculates desired drone position, taking into account the follow target altitude mode
+	 *
```
`@param[in]`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
-			matrix::Vector3f acc_ned_est);
+	/**
+	 * Calculate the gimbal height offset to the target to calculate the pitch angle command
+	 *
```
`@param[in]`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
-	matrix::Vector3f calculate_drone_desired_position(matrix::Vector3f target_position, matrix::Vector2f offset_vector);
+	/**
+	 * Publishes gimbal control command to track the target, given xy distance and z (height) difference.
+	 */
```
`@param[in]`

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	SecondOrderReferenceModel<matrix::Vector3f> _target_pose_filter;
+
+	// Internally tracked Follow Target characteristics, to allow RC control input adjustments
+	float _follow_target_distance{0.0f};
```
is this an ok value to initialize this to?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	// Estimated (Filtered) target orientation setpoint
+	float _target_orientation_rad{0.0f};
+
+	// Orbit angle setpoint, measured in global frame, against the target
```
units

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	float _orbit_angle_setpoint{0.0f};
+
+	// Actual drone to target heading [rad]
+	float _drone_to_target_heading{0.0f};
```
does this need to be an internal state?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	float _drone_to_target_heading{0.0f};
+
+	// Actual orbit angle in relation to the target
+	float _measured_orbit_angle{0.0f};
```
does this need to be an internal state?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
-	// Lowpass filter for smoothing the offset vector and have more dynamic shots when target changes direction
-	AlphaFilter<matrix::Vector2f> _offset_vector_filter;
+	// Actual drone to target 2d position vector
+	Vector2f _drone_to_target_vector{0.0f, 0.0f};
```
does this need to be an internal state?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
-	// Lowpass filter assuming  values 0-1, for avoiding big steps in velocity feedforward
-	AlphaFilter<float> _velocity_ff_scale;
+	// Tracked orbit tangential speed, to compensate for the orbital motion for velocity setpoints
+	Vector2f _orbit_tangential_velocity{0.0f, 0.0f};
```
does this need to be an internal state?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	// Variable to remember the home position's z coordinate, which will be baseline for the position z setpoint
+	float _home_position_z{0.0f};
+
+	float _gimbal_pitch{0};
```
does this need to be an internal variable? and if so.. define it with comment

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
+		FlightTask,
+		(ParamFloat<px4::params::NAV_FT_HT>) _param_nav_ft_ht,
 		(ParamFloat<px4::params::NAV_FT_DST>) _param_nav_ft_dst,
 		(ParamInt<px4::params::NAV_FT_FS>) _param_nav_ft_fs,
 		(ParamInt<px4::params::NAV_FT_ALT_M>) _param_nav_ft_alt_m,
-		(ParamInt<px4::params::NAV_FT_GMB_M>) _param_nav_ft_gmb_m,
-		(ParamInt<px4::params::NAV_FT_DELC>) _param_nav_ft_delc
+		(ParamFloat<px4::params::NAV_FT_YAW_T>) _param_ft_yaw_t
```
since these are all defined locally.. would be great to give the local states more verbose names so that we can actually understand what they are within this flight task. e.g. `_param_follow_target_height` (only with param prefix if the idea is that we want it to be clear this was from the params.. and we have other sources that could change the actually used follow target height)

tstastny replied:
i would get rid of all acronyms, so that the variable is readable in the code without needing to go check the para.c file -- the param import here already tells where the value comes from, now in the flight task we just need something we can understand in the lines of the code. the only reason i would put the prefix param is to distinguish between "param" set value and the actual follow height that we are commanding (which can be adjusted from sticks as well). so there would in principle be two variables of the same name, unless we distinguish between the two.

junwoo091400 replied:
There was this discussion 3 years ago exactly about this issue. And it removed all these 'readable' parameter variable names in the code, and reset them to match exactly the parameter name. Could you review this as well for the final decision? thinking @tstastny @potaito

https://github.com/PX4/PX4-Autopilot/pull/11686

junwoo091400 replied:
Right I agree that using the direct parameter name in the code decreases code readability :yum:

potaito replied:
I have to admit I'm siding with #11686 because it's weird to sometimes be able to CTRL+F certain parameters, and other times not, because the handle is not re-using the parameter name :sweat_smile: 

But your argument is valid.

MaEtUgR replied:
@tstastny Thanks for asking. I totally understand your point and the verbose internal variable names were used for a very long time before it was streamlined. But for years I was wasting minutes every day looking up which parameter is used where and what exact parameter contributes to the code you're reading. I have to say it was such a big improvement when the names were streamlined and finally grepable, I never looked back once. As you can see in the pr this relief was shared by a lot of if not all the core contributors that cared. Correct me if I'm wrong. So I would disadvise reverting this naming strategy.

That said I don't like abbreviations in variable names and I encourage to use descriptive unshortened names. Also, I suggest thinking twice when defining or reviewing parameter names to make them as descriptive as possible given the limitations. Ideally parameter names would not be limited to 16 characters only.

tstastny replied:
thanks for the extra insight @MaEtUgR and @potaito  - i can understand the reasoning. i guess then as we discussed in the dev call we should attack the *heart of the issue .. 16 char limits on param names. I'll follow up with @bkueng offline. @junwoo091400 for now you should then stick with the convention from https://github.com/PX4/PX4-Autopilot/pull/11686

tstastny reviewed (changes requested):
@junwoo091400 - some great progress on this. lots of nice features and improvements. main points i noticed during the review - try to const when possible, let's avoid adding more internal states when we don't need to, and we should be careful not to disable old functionality while putting in new (unless it is intentional), e.g. seems that the manual stick inputs are the only way to change the follow height angle or distance now. please let me know if you need any clarification on any of the review comments. cheers.

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	// The drone will move into position as soon as the target starts moving and its heading becomes known.
-	Vector2f current_drone_heading_2d{cosf(_yaw), -sinf(_yaw)};
+	// We don't command the yawspeed, since only the yaw can be calculated off of the target
+	_yawspeed_setpoint = 0.f;
```
Oh yes, good catch. This will be tracked by the controller, and will lower the performance because it's conflicting with `yaw_setpoint`.

potaito replied:
Set it `NaN` instead.

potaito commented:
Also, the submodule `Tools/flightgear_bridge` was updated, but probably not on purpose, right?

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		update_target_pose_filter(_follow_target_estimator);
 
-		// Set position and velocity setpoints
-		float desired_velocity_ff_scale = 0.0f;  // Used to ramp up velocity feedforward, avoiding harsh jumps in the setpoints
```
Is `desired_velocity_ff_scale`, i.e. blending in the velocity-feedforward no longer required?

tstastny replied:
blending no longer needed - he just sets the feed-forward according to the sum of lever arm velocity and target velocity, then MC position controller handles prioritization when near or far from position setpoint.

tstastny replied:
- target velocity is filtered, so no jumps there.
- lever arm based velocity could make a step whenever orbit angle is changed because we immediately apply max orbiting rate, though @junwoo091400 added a ramp down to zero when close to the setpoint, and the orbit angle command is then no longer changed, so no more orbit rate based lever arm velocity.

A final output filter would solve this :) -- but im not sure it will be too deleterious anyway.. would be just like you change the go-to position, the velocity setpoint will have a step input and gradually decrease to zero when the orbit angle reaches it's commanded value.

@junwoo091400 did you notice anything bad looking from this in your recent flights? we could address if the behavior was poor

junwoo091400 replied:
I haven't tested with the orbital velocity bug fix version, will test today :hand:. Since the previous tests only had the orbital velocity setpoint of 1 m/s magnitude, it isn't a valid test case I believe.

junwoo091400 replied:
orbit angle trajectory generator has solved this problem :) Both acc & vel limited!

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -130,81 +141,123 @@ class FlightTaskAutoFollowTarget : public FlightTask
 		(ParamFloat<px4::params::NAV_FT_DST>) _param_nav_ft_dst,
 		(ParamInt<px4::params::NAV_FT_FS>) _param_nav_ft_fs,
 		(ParamInt<px4::params::NAV_FT_ALT_M>) _param_nav_ft_alt_m,
-		(ParamInt<px4::params::NAV_FT_GMB_M>) _param_nav_ft_gmb_m,
```
Is the gimbal tracking mode setting no longer required? What's the tracking mode now?

potaito reviewed (commented):
Also, these submodules were updated, but probably not on purpose, right?
- `src/drivers/gps/devices`
- `Tools/flightgear_bridge`
- `src/modules/mavlink/mavlink`

I'll take another look at the PR once TJ's remarks were addressed. Otherwise it will become too much for everyone involved :joy:

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	if (PX4_ISFINITE(current_drone_heading_2d(0)) && PX4_ISFINITE(current_drone_heading_2d(1))) {
-		_offset_vector_filter.reset(current_drone_heading_2d);
+	// Update the internally tracaked Follow Target characteristics
+	_follow_angle_rad = math::radians(get_follow_me_angle_setting(_param_nav_ft_fs.get()));
+	_follow_target_distance = _param_nav_ft_dst.get();
+	_follow_target_height = _param_nav_ft_ht.get();
```
Nope, I need to add this. Only thing that's bugging me was that we need to check if the parameter value has been updated, to not accidentally overwrite the RC adjusted value, if there is no Parameter Change. And to detect this 'update', we either :
1. Need to store local copy, and compare the new param value every loop
2. Override the "handleParamterUpdate()" function from Flight Task, to update the internal values accordingly, since this will be called when there's a parameter update system-wide.

However, the no.2 solution has a problem where a change in e.g. EKF2 parameter set, will also trigger the function. Not just about the Follow Target paramters.

I believe the best solution will be to add a function in the Params Class to check if it has been 'updated', and call this for each parameter to check for updates, but it hasn't been implemented in the Params class yet.

what do you think of these solutions?

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	} else {
-		_offset_vector_filter.reset(Vector2f(1.0f, 0.0f));
+	// Save the home position z value, to allow user so that altitude setpoint will be relative to arming position (home) altitude
+	if (_sub_home_position.get().valid_alt) {
```
I will make it so that if there's no valid home position altitude, the mode can't be entered :thinking: . But is there a case where home position doensn't have the valid altitude, is the question.

tstastny replied:
> But is there a case where home position doensn't have the valid altitude, is the question

not sure - @potaito ? anyway i don't like leaving things ambiguous or reliant on knowledge of other system components if possible.

potaito replied:
I'm sure you can find corner cases by taking off in manual without a GPS lock and then acquiring a GPS lock mid-flight for example. No home position for you then. So yes, I would assume that the drone can be flying and enter this flight mode without a valid home position.

It also makes sense to restrict the flight mode to be only allowed with a valid home position. There is a parameter for the tracking mode, and depending on its setting, the home position is required. The parameter can be changed mid-flight and during follow-me. This means there are two options:
1. You allow entering the follow-me flight mode in any case, but fall out of it as soon as home position is not valid and the parameter configuration requires it. 
2. You check the home position validity on flight task activation and reject if there is no home position.

I think the second option is more in line with where the PX4 architecture is headed.

junwoo091400 replied:
I was confused at first regarding what 'home position' actually is, and I guess it does get set as a local position from EKF2 when initialized.

https://github.com/PX4/PX4-Autopilot/blob/dab7b007de17b3449b875735d337149716af80c4/src/modules/commander/Commander.cpp#L1776-L1795

So then the altitude itself can actually be trusted, but it's just that it would shift a bit during flight. So I think it's okay to let user enter the mode without home position valid alt? How much can the altitude actually shift after GPS lock?

But my big questions is why would it even shift? (the local position coordinate system), or is the coordinate system being static and only the home position coordinates are changing? I find this 'home position local position coordinates affected by GPS lock' confusing 😢

junwoo091400 replied:
So the home position's xyz will always be valid, since it's from when the drone arms. But, the global lat, lon and alt can be incorrect. So I have added a condition to not let follow me get activated if there's no valid alt (global altitude)

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	} else {
-		// Define and rotate offset vector based on follow-me perspective setting
-		const float new_follow_angle_rad = math::radians(update_follow_me_angle_setting(_param_nav_ft_fs.get()));
+// Update the target pose filter, knowing that target estimator data is valid (checked in the update() main function).
```
Good point! It makes sense to assume that it will have a valid data coming in if it's not handled explicitly!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-			_follow_angle_rad = new_follow_angle_rad;
+	// Handle Timetout cases
+	if(target_estimator_timeout) {
+		// Reset the Target pose filter if it's state is not finite
+		if (!PX4_ISFINITE(_target_pose_filter.getState()(0)) || !PX4_ISFINITE(_target_pose_filter.getState()(1))
+		|| !PX4_ISFINITE(_target_pose_filter.getState()(2)) || !PX4_ISFINITE(_target_pose_filter.getRate()(0)) ||
+			!PX4_ISFINITE(_target_pose_filter.getRate()(1)) || !PX4_ISFINITE(_target_pose_filter.getRate()(2))) {
+			_target_pose_filter.reset(pos_ned_est, vel_ned_est);
```
Right, good point!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		return;
+	}
+
+	// Only apply Follow height adjustment if height setpoint and current height are within 1 second window
```
Got removed through the rebases, will add it!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	}
+
+	// Only apply Follow Angle adjustment if orbit angle setpoint and current orbit angle are within 1 second window
+	if(fabsf(_measured_orbit_angle - _orbit_angle_setpoint) < FOLLOW_ANGLE_USER_ADJUST_SPEED) {
```
Good point!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+
+	// Only apply Follow Angle adjustment if orbit angle setpoint and current orbit angle are within 1 second window
+	if(fabsf(_measured_orbit_angle - _orbit_angle_setpoint) < FOLLOW_ANGLE_USER_ADJUST_SPEED) {
+		// Roll for changing follow angle. When user commands +Roll (right), angle increases (clockwise)
```
nice!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		_offset_vector_filter.update(Vector2f{offset_x, offset_y});
+	// Depending on the target velocity, freeze or set new target orientatino value
+	if(target_velocity_norm >= TARGET_VELOCITY_DEADZONE_FOR_ORIENTATION_TRACKING) {
+		_target_orientation_rad = atan2f(target_velocity(1), target_velocity(0));
```
By const this method, you mean make it not change internal state variables? And yes, this _target_orientation_rad setting inside this function was absolutely unnecessary! Thanks ;)

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	const float raw_target_orbit_angle = matrix::wrap_pi(target_orientation + follow_angle);
+
+	// Calculate orbit angle error & it's direction
+	const float orbit_angle_error = matrix::wrap_pi(raw_target_orbit_angle - _orbit_angle_setpoint);
```
Nice! I like organizing the methods scope local. Makes it easier to track / maintain!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	// Low-pass filter on target position.
-	_target_position_filter.setParameters(_deltatime, POSITION_FILTER_ALPHA);
+float FlightTaskAutoFollowTarget::calculate_gimbal_height(float target_height) {
+	float gimbal_height{0.0f};
```
I am changing this variable in the switch case under this, so can't make it const

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +66,243 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		// Calculate heading to the target (for Yaw setpoint)
+		_drone_to_target_heading = atan2f(_drone_to_target_vector(1), _drone_to_target_vector(0));
+		// Calculate current orbit angle around the target, taken into consideration in RC Follow Angle Adjustment
+		_measured_orbit_angle = matrix::wrap_pi(_drone_to_target_heading + M_PI_F);
```
It's calculating the atan2f of the target to drone, so it is essentially 'drone to target' + 180 degrees, I have added additional commit explaining this.

I did this since I thought doing atan2f again for just the negative- vector is a waste of computational resource (although it would be more intuitive) : e.g. `atan2f(_target_to_drone_vector(1), _target_to_drone_vector(0))`

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -106,39 +94,32 @@ PARAM_DEFINE_FLOAT(NAV_FT_RS, 0.1f);
+ * @decimal 2
  * @group Follow target
  */
-PARAM_DEFINE_INT32(NAV_FT_DELC, 0);
```
The second order filter should not need the filter delay compensation (due to velocity feedback), I believe (@tstastny).

I also think it's better to minimize the number of 'magic numbers' (e.g. filter compensation time constant here)

junwoo091400 commented on src/modules/flight_mode_manager/tasks/Utility/Sticks.hpp:
```diff
@@ -58,6 +58,11 @@ class Sticks : public ModuleParams
 	const matrix::Vector<float, 4> &getPositionExpo() { return _positions_expo; };
 
+	// Helper functions to get stick values more intuitively
+	float getPitch() { return _positions(0); }
+	float getRoll() { return _positions(1); }
+	float getThrottle() { return -_positions(2); } // Convert Z-axis(down) command to Up-axis frame
+	float getYaw() { return _positions(3); }
```
Fixed by https://github.com/PX4/PX4-Autopilot/pull/19309/commits/8b40c1f2e7442143e435a026771bad991efb788c

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -44,15 +44,15 @@
  * @group Follow target
  */
-PARAM_DEFINE_FLOAT(NAV_MIN_FT_HT, 8.0f);
+PARAM_DEFINE_FLOAT(NAV_FT_HT, 8.0f);
```
So far I avoided renaming existing follow-me parameters. If we are going to rename them, we might as well completely redefine them. For example, we could remove the `NAV_` since it's no longer part of navigator.

If we rename existing parameters, we'll also have to update mavsdk to reflect the renaming.

junwoo091400 replied:
Gotcha, and yes we can remove NAV. And maybe do : FOLLOW_TARGET_*

potaito replied:
`FOLLOW_TARGET_` is way too long, yes :sweat_smile: 

You could swap `NAV_FT` with `FLW_TGT`, or `FLW_ME`.

By the way, I still haven't made up my mind whether we should call it follow-target or follow-me :roll_eyes: 
Follow-me sounds much friendlier.

junwoo091400 replied:
FLW_TGT sounds nice 👍

potaito commented on src/modules/logger/logged_topics.cpp:
```diff
@@ -62,9 +62,9 @@ void LoggedTopics::add_default_topics()
-	add_topic("follow_target_status", 100);
 	add_topic("follow_target", 500);
 	add_topic("follow_target_estimator", 100);
+	add_topic("follow_target_status", 100);
```
Before merging we should probably decrease the logging rate to 200ms or even 400ms. For debugging it was nice to have the high rate.

potaito commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1516,9 +1516,9 @@ void MavlinkReceiver::fill_thrust(float *thrust_body_array, uint8_t vehicle_type
 
 	case MAV_TYPE_VTOL_DUOROTOR:
 	case MAV_TYPE_VTOL_QUADROTOR:
+	case MAV_TYPE_VTOL_FIXEDROTOR:
```
Where did these changes come in?

junwoo091400 replied:
It's cherry picked from here, to fix compile issues due to changes in Mavlink submodule!

https://github.com/PX4/PX4-Autopilot/pull/19375

potaito commented on src/modules/mavlink/streams/HIL_ACTUATOR_CONTROLS.hpp:
```diff
@@ -95,7 +95,7 @@ class MavlinkStreamHILActuatorControls : public MavlinkStream
 				    system_type == MAV_TYPE_VTOL_DUOROTOR ||
 				    system_type == MAV_TYPE_VTOL_QUADROTOR ||
-				    system_type == MAV_TYPE_VTOL_RESERVED2) {
+				    system_type == MAV_TYPE_VTOL_FIXEDROTOR) {
```
Where did these changes come in?

junwoo091400 replied:
It's cherry picked from here, to fix compile issues due to changes in Mavlink submodule!

https://github.com/PX4/PX4-Autopilot/pull/19375

junwoo091400 commented on msg/follow_target_status.msg:
```diff
@@ -1,5 +1,11 @@
+float32[3] pos_est_filtered  # [m/s] Filtered target position used for vehicle's setpoint generation
+float32[3] vel_est_filtered  # [m/s] Filtered Target velocity used for vehicle's setpoint generation
+
+float32 tracked_target_orientation # [rad] Tracked target orientation, which can be rate constrained
```
Terminology wise, I'm still confused but yes it's global (and local) NED frame based heading! (North is 0 orientation angle)

bresch replied:
call it `tracked_target_heading` maybe?

tstastny replied:
`heading` is the angle between North and the `air velocity`
`yaw` is the angle between North and the `x-body-axis`
`course` is the angle between North and the `ground velocity`

I originally proposed `orientation` here - because I didnt know if we would switch back and forth between `course` based operations (as it is now) and `yaw` based operations, as could be if we had feedback from the rover, or car, or whatever. `orientation` is then more general. Also the definition is even now a bit ambiguous in our current implementation, because course is undefined at zero speed. If you were to change it, I would only recommend the word `course`, as that's how it's (mostly) implemented now, and I'm on a crusade to lock down consistent aerospace namings :)

but otherwise -- i think if you define it well in the message definition, can keep it as is.

junwoo091400 replied:
I will opt for course!

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	}
+}
+
+void FlightTaskAutoFollowTarget::update_rc_adjusted_follow_height(float &follow_height) {
```
^agree with @potaito - I also would point out I generally prefer to minimize output params (that is params getting input and then output again) if it is possible to "return" them. Helps to more clearly define the input output flow.

My two cents (opinion) - I'd go for @potaito 's second option above, or input the sticks as const ref and update the follow height state (without inputting as param) - as your method name explicitly states that it is going to change that state. So you aren't lying to anyone, and it's not a side effect.

junwoo091400 replied:
Since 'Sticks' is a must-have consideration for this function, passing it is a good idea!

Also I'd go for the Alessandro's 2nd option since it makes everything more verbose (since the internal variables get set in the main loop), but it is a bit more tedious than your suggestion!

junwoo091400 replied:
Actually, I'd go for your suggested option since it's less tedious 😏

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		Vector2f orbit_tangential_velocity{0.0f, 0.0f};
+
+		// Update the new orbit angle (rate constrained)
+		_tracked_orbit_angle_setpoint_rad = update_orbit_angle(orbit_tangential_velocity, _tracked_target_orientation_rad, _tracked_orbit_angle_setpoint_rad);
```
@potaito i agree with the nitpick ;) my general perspective is that if it is possible to functionalize an operation with clear simple inputs and outputs, i'd return the new value of e.g. an internal state. this way the function is const, and also easily testable. HOWEVER - i also think it is ok, if the function name explicitly says it is e.g. updating class state X, that the class state X does not need to be input, and can directly be modified. the idea is just to make it very easy to know what the function is doing. and not have side effects (i.e. stuff happening that the function didnt tell you it was going to do). 

>I think I'm just nitpicking because some class members are passed as method arguments, but not all

yeah, I'd also be careful about mixing things like this. can add confusion at times. again - the rule of thumb is generally just whatever is easily readable and not lying to the reader.

junwoo091400 replied:
Added the update trajectory part to the function name to be not deceptive 👅

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -123,88 +146,162 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	uint64_t _last_valid_target_estimator_timestamp{0};
+
+	// Second Order Filter to calculate kinematically feasible target position
+	SecondOrderReferenceModel<matrix::Vector3f> _target_pose_filter;
```
this is an accidental abuse of terminology recommended to @junwoo091400 by me (my bad!). but you're right to point out its common definition! probably we should change it.

also drawing a blank right now on a single word that encompasses position and velocity.

what about e.g.
- `_target_position_velocity_filter` to brute force it?
- `_target_motion_filter` or `_target_translational_state_filter` likely less clear

other ideas? :)

potaito replied:
`_target_position_velocity_filter` appeals to me the most :smile:

junwoo091400 replied:
_target_position_velocity_filter is great!

tstastny commented on src/lib/mathlib/math/filter/second_order_reference_model.hpp:
```diff
@@ -0,0 +1,153 @@
+
+		filter_accel_ = state_error * spring_constant_ + rate_error * damping_coefficient_;
+		filter_rate_ = integrate(filter_rate_, filter_accel_, time_step);
+		filter_state_ = integrate(filter_state_, filter_rate_, time_step);
```
@bresch yes - on my list for tonight. ideally we get that one merged in first, then it doesnt appear as diff in this PR

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
 	}
+}
+
+float FlightTaskAutoFollowTarget::update_orbit_angle(Vector2f &orbit_tangential_velocity, const float target_orientation, const float previous_orbit_angle_setpoint)
```
I can create : 
1. generate_orbit_angle_trjaectory(), which returns the orbit angle, and then call
2. calculate_orbit_tangential_velocity, which would return calculation based off of the trajectory generated.

**Since TJ mentioned Silently changing internal variables is bad, I would keep a verbose function name like "update .. trajectory" for the first one because it implies that orbit angle trajectory has been generated (internally).**

But, then it's werid to return the new orbit angle, or should I name it : "genreate_orbit_anble_trajectory_and_get_orbit_angle"? 😃

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	}
+	// Reset the orbit angle trajectory generator and set maximum angular acceleration and rate
+	_orbit_angle_traj_generator.reset(0.f, 0.f, 0.f);
+	_orbit_angle_traj_generator.setMaxJerk(_param_mpc_acc_hor_max.get() / _follow_distance);
```
Done!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		// Set position and velocity setpoints
-		float desired_velocity_ff_scale = 0.0f;  // Used to ramp up velocity feedforward, avoiding harsh jumps in the setpoints
+		// Orbital tangential velocity setpoint calculated from orbit angle calculation
+		Vector2f orbit_tangential_velocity{0.0f, 0.0f};
```
Done!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -63,186 +67,272 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	// The drone will move into position as soon as the target starts moving and its heading becomes known.
-	Vector2f current_drone_heading_2d{cosf(_yaw), -sinf(_yaw)};
+	// We don't command the yawspeed, since only the yaw can be calculated off of the target
+	_yawspeed_setpoint = NAN;
```
Current behaviour is to not set yaw at all, until drone & target's distance is bigger than minimum distance for yaw control, then sets as the raw heading from drone to target as the yaw.

junwoo091400 replied:
I don't believe so :thinking:. But also I thought that the sudden 360 degrees yaw spin that happens when I activate the follow me mode is from this hard setpoint, which is why I believe I want to not set it at all.

Since we don't know where the target is :thinking:

junwoo091400 replied:
So the FlightTask (the parent class) initializes (resets) it to NAN, so it won't be commanding at all. I guess then it makes sense to command the current yaw since commanding NAN is unpredictible? 🤔

potaito replied:
Yes, I would also command the current yaw :+1:

bresch replied:
No need to do it now, but don't forget that you're getting `last_setpoint` as a parameter, meaning that you could initialize with it (if finite) to avoid glitches (see https://github.com/PX4/PX4-Autopilot/blob/3ac8fdbe2901dcd812c8e1944f72765ddff4c022/src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp#L81)

junwoo091400 commented on src/lib/mathlib/math/test/second_order_reference_model_test.cpp:
```diff
@@ -0,0 +1,131 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 ECL Development Team. All rights reserved.
```
This is for you @tstastny

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -123,88 +146,162 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	AlphaFilter<float> _yaw_setpoint_filter;
+
+	// Variable to remember the home position's z coordinate, which will be baseline for the position z setpoint
+	float _home_position_z{0.0f};
```
Added logic in activate function to not allow the task to get activated if there's no valid home position!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -257,50 +347,42 @@ bool FlightTaskAutoFollowTarget::update()
+			_yaw_setpoint = _yaw_setpoint_filter.getState();
 		}
+		// Gimbal setpoint
+		const float gimbal_height = calculate_gimbal_height(target_position_filtered(2));
```
I added the altitude mode parameter for the function to make it clear that it's taking altitude mode into account :) Thanks!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -123,88 +146,162 @@ class FlightTaskAutoFollowTarget : public FlightTask
 	 *
-	 * @param param_nav_ft_fs value of the parameter NAV_FT_FS
-	 * @return Angle [deg] from which the drone should view the target while following it, with zero degrees indicating the target's 12 o'clock
+	 * If the drone is within the an user adjustment error time window away from the distance setpoint,
```
I made it clearer 🤝

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -257,50 +347,42 @@ bool FlightTaskAutoFollowTarget::update()
-
-		// Gimbal setpoint
-		float gimbal_height = 0;
+		follow_target_status.emergency_ascent = emergency_ascent; // Log emergency ascent state
```
Solved this by creating 2 local variable in the beginning :)

junwoo091400 commented:
Some related videos I wanted to share:
1. Upstream Follow-Me behavior
https://www.youtube.com/watch?time_continue=42

2. Current PR's Follow-Me behavior
From March 26th:
https://www.youtube.com/watch?time_continue=107
From March 11th:
https://www.youtube.com/watch?v=myPFVJJkm04&t=19s

junwoo091400 commented:
Few Notes on the current development progress:
1. **Right now, this version is ready to go & to be reviewed for merge.** It fixed all the major tiny problems that was arising from algorithmical details of Follow Target.
2. The last challenge was the 'Overshoot' of the 2nd order reference model problem, which was causing the drone to thinking that target was moving backwards, when in fact target is moving still. By including unfiltered velocity into consideration (which, stays at 0 when we have the overshoot phenomenon) this was fixed.
3. Let's get this merged :mechanical_arm:

junwoo091400 commented:
Squashed & Edited past commits to exclude merge conflict ruins (e.g. >>>>> HEAD).
Ready for review @tstastny @potaito 😄

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	// Wrap orbit angle difference, to get the shortest angle between them
+	if(fabsf(matrix::wrap_pi(measured_orbit_angle - tracked_orbit_angle_setpoint)) < FOLLOW_ANGLE_USER_ADJUST_SPEED * USER_ADJUSTMENT_ERROR_TIME_WINDOW) {
+		// RC Roll stick input for changing follow angle. When user commands RC stick input: +Roll (right), angle increases (clockwise)
+		// Constrain adjust speed [rad/s], so that drone can actually catch up. Otherwise, the follow angle
```
smart

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		return;
+	}
+	// Only apply Follow height adjustment if height setpoint and current height are within time window
+	if(fabsf(_position_setpoint(2) - _position(2)) < FOLLOW_HEIGHT_USER_ADJUST_SPEED * USER_ADJUSTMENT_ERROR_TIME_WINDOW) {
```
I don't quite understand why `USER_ADJUSTMENT_ERROR_TIME_WINDOW` is called a time window.

- On the left hand side we have `fabsf(_position_setpoint(2) - _position(2))`, which is the tracking error.
- On the right hand side we have two constant variables acting as a threshold.

Does this prevent adjustments of heigh / angle / distance, when the error is greater than what's deemed an acceptable error? Why is this necessary? I mean, what real situation is it preventing from happening?

junwoo091400 replied:
It's well described in the follow angle adjustment [function](https://github.com/PX4/PX4-Autopilot/pull/19260/files#diff-c427e5824011523926ec83abd52b8dfe31261a95bf66c1b12394fb79573d84fdR169) below.

If the follow characteristics are changed without taking real drone's position into consideration, the drone can end up following that 'imaginary' setpoint and user's RC input won't actually affect drone's behavior since the setpoint is so far fetched (especially the case for angle adjustments)

potaito replied:
Interesting, now I understand why you're doing it! 
But where does "time" come into play? I'm just hung up on the terminology.

junwoo091400 replied:
I found it hard to generalize the calculation of an arbitrary constraint for follow distance, angle and height.

Which is why I went with a time constant, since then it can generally provide enough buffer zone in user's perspective (max 0.5 second delay from control input to action of the drone, roughly. Does that make sense?

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -37,63 +37,94 @@
+
+// [m/s] Speed with which the follow distance will be adjusted by when commanded with deflection via RC command
+static constexpr float FOLLOW_DISTANCE_USER_ADJUST_SPEED = 2.0;
+// [m] Maximum follow distance that can be set by user's RC adjustment
```
I see the sentiment here where you wanted to group similar entries together by omitting a new line in between them. Personally though I think it lowers the readability, because it's an irregularity.

junwoo091400 replied:
Sure! :)

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -37,63 +37,94 @@
+
+// [m/s] Speed with which the follow height will be adjusted by when commanded with deflection via RC command
+static constexpr float FOLLOW_HEIGHT_USER_ADJUST_SPEED = 1.5;
+// [m] Maximum follow height that can be set by user's RC adjustment
```
Same here. Personally I would add a blank line in between.

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -37,63 +37,94 @@
+// Prevents setpoints diverging from the vehicle's actual position too much
+static constexpr float USER_ADJUSTMENT_ERROR_TIME_WINDOW = 0.5f;
+
+// Deadzone on both +/- direction for normalized stick input (-1, +1) where user adjustment will be ignored
```
I wonder if the deadzone is necessary. It probably does not hurt to have some minimal deadzone. But why does the orbit flight task not require one too? Because orbit is using an exponential of the stick value, thus mitigating noise around 0?

junwoo091400 replied:
That's true. I could def use the Expo! I added this since the follow angle adjustment was really sensitive (especially when I just want to change the follow distance : pitch stick), it would swivel around the follow angle (roll stick) quite a lot.

But Expo could help fix this!

junwoo091400 replied:
I would do a separate PR on adding new verbose functions for expo values as well (currently it's index based, and I would like to add : getPitchExpo(), and so on). Or, should I include it to this PR?

potaito replied:
Sounds good

junwoo091400 replied:
Done!

junwoo091400 replied:
https://github.com/PX4/PX4-Autopilot/pull/19591

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +161,189 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	/**
+	 * Calculate the tracked target orientation and overwrite the tracked target orientation if necessary
+	 *
+	 * Note : Filtered target velocity is generated via 2nd order reference model filter can have overshooting behaviors
```
Is there a link to any literature regarding "2nd order reference model"? I don't think this is a universal expression known to most developers.

tstastny replied:
you could say something like "the target velocity is filtered through a second order multi-input multi-output system (reference model) and may overshoot when the natural frequency of the system is set to a low value" -- TMI? :)

junwoo091400 replied:
Still not sure about this, is MIMO more well known term for developers? Otherwise, I would just mention the library header location to say : "If you don't know about this reference model, check out this library"

tstastny replied:
probably easiest is just to say "second order filter". It's just a second order dynamic system we use as a filter. I don't think you need to explain it in detail here. the reader just knows it is a filter, and can look up the implementation. aerospace controls people probably understand the term "reference model" more than others.

junwoo091400 replied:
Alrighty :) I opted for 'second order filter'

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +161,189 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * Release Gimbal Control
+	 *
+	 * Releases Gimbal Control Authority of Follow-Target Flight Task, to allow other modules / Ground station
+	 * to control the gimbal when the task exits.
```
```suggestion
	 * to control the gimbal when the task exits.
	 * Fore more information on gimbal v2, see https://mavlink.io/en/services/gimbal_v2.html
```

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -106,39 +108,22 @@ PARAM_DEFINE_FLOAT(NAV_FT_RS, 0.1f);
- *
- * @value 0 Maintain constant altitude and track XY position only (2D tracking)
- * @value 1 Track target's altitude (3D tracking)
+ * @value 0 2D Tracking : Maintain constant altitude relative to home and track XY position only
```
```suggestion
 * @value 0 2D Tracking: Maintain constant altitude relative to home and track XY position only
```

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -106,39 +108,22 @@ PARAM_DEFINE_FLOAT(NAV_FT_RS, 0.1f);
- * @value 0 Maintain constant altitude and track XY position only (2D tracking)
- * @value 1 Track target's altitude (3D tracking)
+ * @value 0 2D Tracking : Maintain constant altitude relative to home and track XY position only
+ * @value 1 2D + Terrain : Mantain constant altitude relative to terrain below and track XY position
```
```suggestion
 * @value 1 2D + Terrain: Mantain constant altitude relative to terrain below and track XY position
```

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -106,39 +108,22 @@ PARAM_DEFINE_FLOAT(NAV_FT_RS, 0.1f);
- * @value 1 Track target's altitude (3D tracking)
+ * @value 0 2D Tracking : Maintain constant altitude relative to home and track XY position only
+ * @value 1 2D + Terrain : Mantain constant altitude relative to terrain below and track XY position
+ * @value 2 3D Tracking : Track target's altitude (be aware that GPS altitude bias usually makes this useless)
```
```suggestion
 * @value 2 3D Tracking: Track target's altitude (be aware that GPS altitude bias usually makes this useless)
```

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -106,39 +108,22 @@ PARAM_DEFINE_FLOAT(NAV_FT_RS, 0.1f);
- *
- * @value 0 disabled
- * @value 1 enabled
+ * This is the maximum velocity the drone will circle around the target whenever
```
This is the "**tangential** velocity" on the circle then, right? Not angular velocity.

potaito reviewed (changes requested):
Nice, looks really good!

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		} else {
-			_follow_angle_rad = new_follow_angle_rad;
-		}
+	// Check Follow target estimator's validity & timeout conditions
```
```suggestion
```
That comment is just repeating what's already explicitly written in code

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-			_follow_angle_filter.reset(_follow_angle_filter.getState() + M_TWOPI_F);
-			_follow_angle_rad = _follow_angle_rad + M_TWOPI_F;
-		}
+	// Reset last valid estimator data received timestamp
```
```suggestion
```

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-			// vieweing angle of the drone
-			_target_velocity_unit_vector = Vector2f(vel_ned_est.xy()).unit_or_zero();
-		}
+	// Reset the filter if any pose filter states are not finite or estimator timed out
```
```suggestion
```

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	// Offset from the Target
+	const Vector2f offset_vector = Vector2f(cosf(orbit_angle_setpoint), sinf(orbit_angle_setpoint)) * _follow_distance;
+
+	// Calculate desired 2D position
```
it's already named `drone_desired_position`
```suggestion
```

bresch reviewed (changes requested):
Looks good, I'm just against redundant comments.
As Uncle Bob said: "_The proper use of comments is: to compensate for our failure to express ourselves in code._" (ref https://bpoplauschi.github.io/2021/01/20/Clean-Code-Comments-by-Uncle-Bob-part-2.html)

junwoo091400 commented on src/modules/commander/Commander.cpp:
```diff
@@ -3489,6 +3488,9 @@ Commander::update_control_mode()
 
 		break;
 
+	// Follow Target supports RC adjustment, so disable auto control mode, which forces
```
Done, but why?

potaito replied:
Because everything before `NAVIGATION_STATE_AUTO_FOLLOW_TARGET:` belongs to the previous switch case :)

junwoo091400 replied:
Hmmm I thought after the break; it's the next entry's zone but I guess that also makes sense :thinking:

potaito commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,13 +43,17 @@
-FlightTaskAutoFollowTarget::FlightTaskAutoFollowTarget()
+// Call Sticks constructor to set Follow Target class as a parent to Sticks object,
+// which enables chained paramter update sequence, defined in ModuleParams
+// to keep the parameters in Sticks up to date.
```
But if I make a comment like this it means, that it wasn't clear to me :sweat_smile: 
Don't just close the comments then :sweat:

junwoo091400 replied:
Ooh I thought you were suggesting a change to the comments, okay let me know if the updated comment is clearer! :pray:

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+		// RC Roll stick input for changing follow angle. When user commands RC stick input: +Roll (right), angle increases (clockwise)
+		// Constrain adjust speed [rad/s] so that drone can actually catch up. Otherwise, the follow angle
+		// command can be too ahead that drone's behavior would be un-responsive to RC stick inputs.
+		const float angle_adjust_speed_max = min(FOLLOW_ANGLE_USER_ADJUST_SPEED, _param_mpc_xy_vel_max.get() / _follow_distance);
```
TODO : constrain on 'Orbit Tangential Vel limit param', to not use MPC's param

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * @param target_position Tracked target position Vector3f reference
+	 * @param orbit_angle_setpoint Current orbit angle setpoint around the target
+	 *
+	 * @return Position [m,m,m] Final position setpoint for the drone
```
one `[m]` is enough i think, functions shouldnt be returning mixed unit vectors.. -- however what would be nice is to state what frame the output is in. e.g. `(NED)`

junwoo091400 replied:
Frames were added!

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * Calculates desired drone position taking into account orbit angle and the follow target altitude mode
+	 *
+	 * @param target_position Tracked target position Vector3f reference
+	 * @param orbit_angle_setpoint Current orbit angle setpoint around the target
```
```suggestion
	 * @param orbit_angle_setpoint [rad] Current orbit angle setpoint around the target
```

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	Vector3f calculate_desired_drone_position(const Vector3f &target_position, const float orbit_angle_setpoint) const;
+
+	/**
+	 * Calculate the gimbal height offset to the target to calculate the pitch angle command
```
```suggestion
	 * Calculate the gimbal height offset to the target
```
it is not necessary that the return be used for pitch considerations, just talk about what *this method does.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * Calculate the gimbal height offset to the target to calculate the pitch angle command
+	 *
+	 * @param altitude_mode Current Follow Target Altitude mode
+	 * @param target_pos_z Target's local position z value to use for 3D tracking
```
```suggestion
	 * @param target_pos_z [m] Target's local position z value to use for 3D tracking
```

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	/**
+	 * Publishes gimbal control command to track the target, given xy distance and z (height) difference
+	 *
+	 * @param xy_distance Horizontal distance to target
+	 * @param z_distance Vertical distance to target
```
units

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * @param follow_perspective value of the parameter FLW_TGT_FS
+	 * @return Angle [deg] from which the drone should view the target while following it, with zero degrees indicating the target's 12 o'clock
+	 */
+	float get_follow_angle_setting_deg(const FollowPerspective follow_perspective) const;
```
see previous comment.. i think this method can be removed

junwoo091400 replied:
Yup!

junwoo091400 replied:
Ooh but I think that if a user sets an invalid parameter, it could access the wrong index of the array of angles. Therefore having a safety net function to catch cases like this would be better in my opinion.

tstastny replied:
what about something like this?
```
float convertFollowPerspectiveToRadians(const FollowPerspective follow_perspective) const
{
	const float follow_angle_rad = (follow_perspective < kFollowPerspectiveInvalid) ? _follow_perspective_angles[follow_perspective] : kFollowPerspectiveBehind;
	return math::radians(follow_angle_rad);
}
```
where, in class declaration:
```
static constexpr float kFollowPerspectiveAnglesDeg[xx] = {0.0f, 45.0f, ....};
```

junwoo091400 replied:
Yes I thought the function name is a bit too vague, I like this better!

tstastny reviewed (changes requested):
Looking really good @junwoo091400 ! A few more comments from my side.. but mostly just typos, rewordings, or forgotten units.

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/follow_target_params.c:
```diff
@@ -37,22 +37,32 @@
+ * Follow target height
  *
- * The minimum height in meters relative to home for following a target
+ * The height in meters for following a target
```
Yes @unit is included already. And it would be the relative height (depending on which altitude mode we use of course, which would assume different height for the target).

tstastny replied:
you can just specific the unit in the tag. but the various behavior of the parameter should be described

junwoo091400 replied:
Ahaaa you mean the comment that says 'meters', yes that's unnecessary!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		} else {
-			_follow_angle_rad = new_follow_angle_rad;
-		}
+	const bool target_estimator_timeout = ((follow_target_estimator.timestamp - _last_valid_target_estimator_timestamp) > TARGET_ESTIMATOR_TIMEOUT_US);
```
Right!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
 		FOLLOW_PERSPECTIVE_BEHIND_RIGHT_ANGLE_DEG = 135,
-		FOLLOW_PERSPECTIVE_BEHIND_LEFT_ANGLE_DEG = 225
+		FOLLOW_PERSPECTIVE_BEHIND_LEFT_ANGLE_DEG = -135
 	};
```
Makes sense! Right it's hard to call it an enum when it doesn't enumerate :chicken:

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	/**
+	 * Calculates desired drone position taking into account orbit angle and the follow target altitude mode
+	 *
+	 * @param target_position Tracked target position Vector3f reference
```
Correct!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	void update_target_orientation(float &current_target_orientation, const Vector2f &target_velocity, const Vector2f &target_velocity_unfiltered);
+
+	/**
+	 * Updates the orbit angle setpoint, taking into account the maximal orbit tangential speed
+	 *
+	 * Returns the orbit angle setpoint, taking into account the maximal orbit tangential speed.
+	 * While setting the orbital velocity setpoint vector for the according position setpoint
```
Yes the comment was outdated! this is better :wink:

junwoo091400 commented on msg/follow_target_status.msg:
```diff
@@ -1,5 +1,18 @@
+float32 orbit_angle_setpoint      # [rad] Current orbit angle setpoint from the smooth trajectory generator
+float32 raw_orbit_angle_setpoint  # [rad] Raw 'idealistic' orbit angle setpoint (has discrete jumps)
+
+bool emergency_ascent             # [bool] True when doing emergency ascent (when distance to ground is below safety altitude)
```
Better! Changes applied

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +68,366 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	// The drone will move into position as soon as the target starts moving and its heading becomes known.
-	Vector2f current_drone_heading_2d{cosf(_yaw), -sinf(_yaw)};
+	// We don't command the yawspeed, since only the yaw can be calculated off of the target
+	_yawspeed_setpoint = NAN;
```
Changes applied!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		_offset_vector_filter.update(Vector2f{offset_x, offset_y});
+void FlightTaskAutoFollowTarget::update_target_orientation(float &current_target_orientation, const Vector2f &target_velocity, const Vector2f &target_velocity_unfiltered)
+{
+	// If both filtered & unfiltered target velocity are above the deadzone velocity, update the target orientation
```
This description is already in the [header](https://github.com/PX4/PX4-Autopilot/pull/19260/files#diff-f38015eaf341625ca8db80ae8b79a9d98f0fe1da6625adc96a12b8e19e6d5da7R225-R228), would it be necessary to include it in both header and source then?

tstastny replied:
i would generally add implementation notes where they occur in the definition. header declaration is generally just for summary of the overall functionality.

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+	// Calculate trajectory towards the unwrapped target orbit angle
+	_orbit_angle_traj_generator.updateTraj(_deltatime);
+
+	// Get the calculaed angle setpoint
```
Removed!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
+
 bool FlightTaskAutoFollowTarget::update()
 {
+	follow_target_status_s follow_target_status{}; // Debugging uORB message for follow target
+	// Members of the uORB message that gets set in the code below
```
Agreed!

junwoo091400 replied:
Maybe it's just that I recently wrote a blog post on uORB so I am trying to be as user friendly as possible (in case someone doesn't know that follow_target_s struct implies that it's an uORB message)

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
 
-		if (_emergency_ascent) {
+		// Emergency ascent when too close to the ground
+		emergency_ascent = PX4_ISFINITE(_dist_to_ground) && _dist_to_ground < MINIMUM_SAFETY_ALTITUDE;
```
renamed!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-
-		point_gimbal_at(target_to_drone_xy.norm(), gimbal_height);
-
 	} else {
 		// Control setpoint: Stay in current position
 		_position_setpoint(0) = _position_setpoint(1) = NAN;
 		_velocity_setpoint.xy() = 0;
 	}
```
Changed to apply NAN for all position setpoints & 0 for all velocity setpoints

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -37,63 +37,96 @@
+// << Target Position Velocity Estimator related constants >>
+
+// [rad/s] Second Order reference model filter natural frequency
+static constexpr float TARGET_POS_VEL_FILTER_NATURAL_FREQUENCY = 1.0f;
```
2 rad/s had some jumpy behaviors, I could check more using the script, but 1 rad/s was definitely never causing problem so I went for this.

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -37,63 +37,96 @@
+// [.] Second Order reference model filter damping ratio
+static constexpr float TARGET_POS_VEL_FILTER_DAMPING_RATIO = 0.7071;
+
+// [us] If the target estimator output isn't updated longer than this, reset pose filter.
```
Changed comment to say pos/vel!

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -105,8 +138,13 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	// Override parameter update function to check when Follow Target properties are changed
+	void updateParams() override;
+
 protected:
```
I am not sure exactly what the benefit of using underscore postfix would be:

"My C# team use underscore prefix for private members, as it works very well with Intellisense because while coding you can type underscore and filter out everything but private members."
https://softwareengineering.stackexchange.com/questions/204443/coding-convention-regarding-the-usage-of-underscores

What benefit would the postfix bring, instead of using the suffix?

tstastny replied:
there are arguments in both directions -- e.g. in the article you linked:
>From a c/c++ point of view
Prefix underscores are only safe with lower case variables, and even then double underscores are reserved
Suffix underscores are always safe, but lead to very ugly constructions like member_->pointer or member_[3]

the point here is that we pick a style and keep it consistent. we chose google style, because it is well documented, has many automated tools available already, is quite close to the majority of most recent PX4 activity, and requires less effort (not needing to re-create an entire style guide and go down the rabbit hole of every user's vastly differing opinions on style)

junwoo091400 commented:
I cherry-picked #19591 into the PR and applied the usage of Expo() values for User RC adjustments.

Also, addressed all the comments! Is it ready to merge?

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -64,306 +70,355 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-		// No or invalid option
-		break;
+	if (follow_perspective < 0 || follow_perspective >= kFollowPerspectiveInvalid) {
+		return math::radians(kFollowPerspectiveAnglesDeg[kFollowPerspectiveInvalid]);
```
eep! this will access something out of the bounds of the array. Maybe choose an existing follow perspective to default? e.g. Back? Whatever you like the best.

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -105,111 +135,212 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * follow_distance will be adjusted with a speed proportional to user RC command
+	 *
+	 * @param sticks Sticks object to get RC commanded values for adjustments
+	 * @param drone_to_target_vector Tracked follow distance variable reference which will be updated to the new value
```
units

tstastny commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -105,111 +135,212 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * overshoot, the unfiltered velocity stays close to 0 (indicating target already stopped), therefore not triggering a target orientation setting.
+	 *
+	 * @param current_target_orientation [rad] Tracked target orientation value that will be over-written with new orientation
+	 * @param target_velocity [m/s, m/s] Filtered Target velocity from which we will calculate target orientation
+	 * @param target_velocity_unfiltered [m/s, m/s] Unfiltered Target velocity that aids in verifying if filtered velocity is accurate
+	 */
+	void update_target_orientation(float &current_target_orientation, const Vector2f &target_velocity, const Vector2f &target_velocity_unfiltered) const;
```
this comment is still not addressed https://github.com/PX4/PX4-Autopilot/pull/19260#discussion_r862709971

1. more intuitive to *return* the target orientation (not do param[in,out])
2. still defining [m/s, m/s] unnecessarily:

tstastny reviewed (changes requested):
@junwoo091400 almost!

A few comments below -- also the style comment was also still not addressed. The method name styles are inconsistent. Should switch them all to lowerCamelCase like the flight task base class and your newly implemented function shown here:
https://github.com/PX4/PX4-Autopilot/blob/20355e8c18515a7b53f52ad6407b742ee042a73e/src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp#L257-L267

junwoo091400 commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -125,91 +163,190 @@ class FlightTaskAutoFollowTarget : public FlightTask
+	 * To check if the filtered velocity is staying true to target's actual motion, unfiltered velocity needs to be taken into account. Since during
+	 * overshoot, the unfiltered velocity stays close to 0 (indicating target already stopped), therefore not triggering a target orientation setting.
+	 *
+	 * @param current_target_orientation  Tracked target orientation value that will be over-written with new orientation
```
Thanks for the valuable feedback, I missed this detail!

junwoo091400 commented:
Addressed all the comments! @tstastny

junwoo091400 commented:
Force pushed to confine FollowPerspective enum to uint8_t type to avoid followperspective being casted to negative value.
https://en.wikipedia.org/wiki/C%2B%2B11#Strongly_typed_enumerations

potaito reviewed (approved):
LGTM :clap:

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,330 +43,386 @@
-		// is necessary because the yaw angle is run through a low-pass filter.
-		if (_follow_angle_rad - new_follow_angle_rad  > M_PI_F) {
-			_follow_angle_rad = new_follow_angle_rad + M_TWOPI_F;
+	// Compare param values and if they have changed, update the properties
+	if (matrix::isEqualF(follow_distance_prev, _param_flw_tgt_dst.get())) {
```
That statement doesn't do what the comment says...

junwoo091400 replied:
Oh it is definitely a bug!!! :palm_tree: How did I miss this!

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,330 +43,386 @@
+	if (matrix::isEqualF(follow_distance_prev, _param_flw_tgt_dst.get())) {
+		_follow_distance = _param_flw_tgt_dst.get();
+	}
+	if (matrix::isEqualF(follow_height_prev, _param_flw_tgt_ht.get())) {
```
same here

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,330 +43,386 @@
+		const Vector2f drone_to_target_vector  = Vector2f(target_position_filtered.xy()) - Vector2f(_position.xy());
+
+		// Calculate heading to the target (for Yaw setpoint)
+		const float drone_to_target_heading = atan2f(drone_to_target_vector(1), drone_to_target_vector(0));
```
Something to test, but you might get a more stable yaw setpoint if you use the heading from the position setpoint to the target estimate. The drone should never be far from the position setpoint anyway.

junwoo091400 replied:
Yes and no, I think the yaw control will be smoother but when the target orientation changes rapidly for example, I think it could lag quite a lot and degrade capability of tracking the target in the center. Something to test!

junwoo091400 replied:
Added to the notes in the PR!

junwoo091400 replied:
I added my test result & thoughts in the [comment ](https://github.com/PX4/PX4-Autopilot/pull/19260#issuecomment-1122821572)below. My conclusion is that keeping it as-is is better!

bresch replied:
Thanks for the testing, I agree with your conclusion

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,330 +43,386 @@
 
-		// Set position and velocity setpoints
-		float desired_velocity_ff_scale = 0.0f;  // Used to ramp up velocity feedforward, avoiding harsh jumps in the setpoints
+		// Update target orientation to track
+		_target_course_rad = updateTargetOrientation(_target_course_rad, target_velocity_filtered.xy(), Vector3f(_follow_target_estimator.vel_est).xy());
```
The comment is useless. Rename `updateTargetOrientation` -> `updateTargetOrientationToTrack` if you think the function name isn't clear enough (I don't think it's useful).

junwoo091400 replied:
Comment removed!

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp:
```diff
@@ -43,330 +43,386 @@
+	drone_desired_position_raw.copyTo(follow_target_status.desired_position_raw);
 
-		point_gimbal_at(target_to_drone_xy.norm(), gimbal_height);
+	// Etc
```
Etc?

junwoo091400 replied:
Removed, it was meant to say : "Other things to log .."

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -105,111 +135,214 @@ class FlightTaskAutoFollowTarget : public FlightTask
+		(ParamFloat<px4::params::FLW_TGT_MAX_VEL>) _param_flw_tgt_max_vel
 	)
 
+	// uORB topics to subscribe to
```
It's called `uORB::Subscription`

bresch commented on src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.hpp:
```diff
@@ -105,111 +135,214 @@ class FlightTaskAutoFollowTarget : public FlightTask
 
-	uORB::PublicationMulti<follow_target_status_s> _follow_target_status_pub{ORB_ID(follow_target_status)};
-	uORB::PublicationMulti<gimbal_manager_set_attitude_s> _gimbal_manager_set_attitude_pub{ORB_ID(gimbal_manager_set_attitude)};
+	// uORB topics to publish
```
`uORB::Publication`, I guess it's clear enough

junwoo091400 replied:
Yep

bresch reviewed (changes requested):
It's mostly good to me, just a small mistake on the param update.
There are still way too many comments IMO, but it's not a big problem.

junwoo091400 commented:
![image](https://user-images.githubusercontent.com/23277211/167713471-67f9e76a-3008-43bd-8238-6d49d5126448.png)
Left : With yaw setpoint as (Position Setpoint -> Target) | Right : With yaw setpoint as (Actual Position -> Target)

@bresch I didn't find significant difference in the yaw / yaw rate command difference in between two different yaw setpoint calculation methods.

However, I did find that basing the yaw calculation off of position setpoint can introduce a yaw tracking lag when there's a difference between the setpoint & actual position:

![Yaw_delay_demonstration](https://user-images.githubusercontent.com/23277211/167714302-ace5cf33-5b6e-4e53-9a63-6ce09c1d8d77.JPG)

Here, it is visible that the PS (Position Setpoint) is leading the AP (Actual Position), and target is moving in upwards-right direction. The yaw setpoint (PS -> Tgt) is calculated as to point lower-right direction, but the correct yaw setpoint (AP -> Tgt) should point right direction. Therefore, the camera ends up pointing on the trailing side of the target. Which is why in the video I am almost out of frame.

I think keeping the yaw setpoint based on Actual Position makes sense in this regards. What do you think? @bresch

junwoo091400 commented:
https://youtu.be/C1Jy3ln028M?t=143

This is the video link to where the yaw legging behavior can be observed (timestamp included)!

junwoo091400 commented:
Oh and yesterday I also tested RC adjustments & Whether changing the Parameter resets the Follow distance, angle and height, it all worked :+1:
